### PR TITLE
Added pointcloud ordered by Z rendering for 2d views

### PR DESCRIPTION
**Pull request by uclaros**
## Description
This PR adds an option to have pointclouds rendered in Z order on 2d map canvas. 

With bottom to top ordering enabled, points with larger Z values cover lower points giving the looks of a true ortho photo. There's also an option for reverse sorting - Top to bottom - where the scene appears as viewed from below.

image

This is how it looks on a sample dataset compared to default rendering.

Default:
image

Bottom to top:
image

### CODE REVIEW

**Comment by Jean-Roc** [comment]
Would it be pertinent to make "top to bottom" the default behavior ?

The input may not be correctly ordered (and I don't know if the untwine stage does any reordering) but the most frequent usecases covering aerial acquisitions, the user will expect to see the roof's points over the ground points.

**Comment by uclaros** [comment]
While for most use cases the user expects to see a top view just like the Bottom to top ordered scene appears, the Default ordering offers the significant advantage of partially rendering the pointcloud as each node is parsed. For the ordered rendering the user needs to wait for all nodes to be parsed and sorted before rendering starts, which admittedly is not the best UX...

**Comment by roya0045** [comment]
> While for most use cases the user expects to see a top view just like the Bottom to top ordered scene appears, the Default ordering offers the significant advantage of partially rendering the pointcloud as each node is parsed. For the ordered rendering the user needs to wait for all nodes to be parsed and sorted before rendering starts, which admittedly is not the best UX...

I'm not too familiar with the internals, but would it be possible to avoid parsing anything and just filter things? What I'm thinking of is, if a point occupies the same pixel as another, don't render it if the value is greater or smaller (depending on filter). Would such a thing be doable or is parsing everything necessary ? I also don't know if it's something that is done once and kept in memory or done for each render.

**Review by wonder-sk — COMMENTED** [review]
Looks good to me, just minor comments

**Comment by uclaros** [comment]
> What I'm thinking of is, if a point occupies the same pixel as another, don't render it if the value is greater or smaller (depending on filter). Would such a thing be doable or is parsing everything necessary ?

That would not be easy with partial overlaps, and especially with antialiased circles.

**Review by wonder-sk — APPROVED** [review]

**Comment by Jean-Roc** [comment]
> For the ordered rendering the user needs to wait for all nodes to be parsed and sorted before rendering starts, which admittedly is not the best UX...

It may not be the best UX or within the scope of this PR but it is often the expected behiavor, be it for an aerial lidar or a subterranean survey, all the point clouds viewing tools we use are displaying data sorted by z dimension, even when using another dimension as the active style (class, intensity). This feature is great as it offers the possibility but it relies on the end-user to discover it.

**Comment by wonder-sk** [comment]
Unrelated test failures.

@Jean-Roc I see your point, but for the time being I would stick to conservative option and keep the original behavior - also because of the fact that ordered rendering has higher CPU and memory requirements. It's easy to modify the default later if needed...

**Comment by wonder-sk** [comment]
This PR is ready to be merged, just the broken CI blocks merging for 3.24...

File: src/core/pointcloud/qgspointcloudlayerrenderer.cpp [hunk 1]
```diff
@@ -116,11 +116,14 @@ bool QgsPointCloudLayerRenderer::render()
   mAttributes.push_back( QgsPointCloudAttribute( QStringLiteral( "X" ), QgsPointCloudAttribute::Int32 ) );
   mAttributes.push_back( QgsPointCloudAttribute( QStringLiteral( "Y" ), QgsPointCloudAttribute::Int32 ) );
 
+  if ( !context.renderContext().zRange().isInfinite() ||
+       mRenderer->drawOrder2d() == QgsPointCloudRenderer::DrawOrder::BottomToTop ||
+       mRenderer->drawOrder2d() == QgsPointCloudRenderer::DrawOrder::TopToBottom )
+    mAttributes.push_back( QgsPointCloudAttribute( QStringLiteral( "Z" ), QgsPointCloudAttribute::Int32 ) );
```
Comment (nyalldawson) [inline src/core/pointcloud/qgspointcloudlayerrenderer.cpp:122]: better to use a switch here -- if we added additional draw orders in future then we'd want a warning raised here

File: src/core/pointcloud/qgspointcloudlayerrenderer.cpp [hunk 2]
```diff
@@ -179,13 +182,25 @@ bool QgsPointCloudLayerRenderer::render()
...
+    }
+    case QgsPointCloudRenderer::DrawOrder::Default:
+    {
+      if ( pc->accessType() == QgsPointCloudIndex::AccessType::Local )
+      {
+        nodesDrawn += renderNodesSync( nodes, pc, context, request, canceled );
+      }
+      else if ( pc->accessType() == QgsPointCloudIndex::AccessType::Remote )
+      {
+        nodesDrawn += renderNodesAsync( nodes, pc, context, request, canceled );
```
Comment (nyalldawson) [inline src/core/pointcloud/qgspointcloudlayerrenderer.cpp:201]: use switch

File: src/core/pointcloud/qgspointcloudlayerrenderer.cpp [hunk 3]
```diff
@@ -342,6 +357,124 @@ int QgsPointCloudLayerRenderer::renderNodesAsync( const QVector<IndexedPointClou
...
+    ++blockCount;
+  }
+
+  if ( pointCount == 0 )
+    return 0;
+
+  if ( order == QgsPointCloudRenderer::DrawOrder::BottomToTop )
+    std::sort( allPairs.begin(), allPairs.end(), []( QPair<int, double> a, QPair<int, double> b ) { return a.second < b.second; } );
+  else if ( order == QgsPointCloudRenderer::DrawOrder::TopToBottom )
+    std::sort( allPairs.begin(), allPairs.end(), []( QPair<int, double> a, QPair<int, double> b ) { return a.second > b.second; } );
```
Comment (nyalldawson) [inline src/core/pointcloud/qgspointcloudlayerrenderer.cpp:450]: use switch

File: src/core/pointcloud/qgspointcloudlayerrenderer.cpp [hunk 4]
```diff
@@ -342,6 +357,124 @@ int QgsPointCloudLayerRenderer::renderNodesAsync( const QVector<IndexedPointClou
...
+
+  // Now we can reconstruct a byte array sorted by Z value
+  QByteArray sortedByteArray;
+  for ( QPair<int, double> pair : allPairs )
```
Comment (nyalldawson) [inline src/core/pointcloud/qgspointcloudlayerrenderer.cpp:454]:
```suggestion
  QByteArray sortedByteArray;
  sortedByteArray.reserve( allPairs.size() );
  for ( QPair<int, double> pair : allPairs )
```

File: src/core/pointcloud/qgspointcloudrenderer.h [hunk 5]
```diff
@@ -272,6 +272,17 @@ class CORE_EXPORT QgsPointCloudRenderer
       Circle, //!< Renders points as circles
     };
 
+    /**
+     * Pointcloud rendering order for 2d views
+     * /since QGIS 3.24
+     */
+    enum DrawOrder
```
Comment (nyalldawson) [inline src/core/pointcloud/qgspointcloudrenderer.h:279]:
```suggestion
    enum class DrawOrder : int
```

File: src/gui/pointcloud/qgspointcloudrendererpropertieswidget.cpp [hunk 6]
```diff
@@ -101,6 +101,10 @@ QgsPointCloudRendererPropertiesWidget::QgsPointCloudRendererPropertiesWidget( Qg
   connect( mPointSizeSpinBox, qOverload<double>( &QgsDoubleSpinBox::valueChanged ), this, &QgsPointCloudRendererPropertiesWidget::emitWidgetChanged );
   connect( mPointSizeUnitWidget, &QgsUnitSelectionWidget::changed, this, &QgsPointCloudRendererPropertiesWidget::emitWidgetChanged );
 
+  mDrawOrderComboBox->addItem( tr( "Default" ), QgsPointCloudRenderer::DrawOrder::Default );
+  mDrawOrderComboBox->addItem( tr( "Bottom to top" ), QgsPointCloudRenderer::DrawOrder::BottomToTop );
+  mDrawOrderComboBox->addItem( tr( "Top to bottom" ), QgsPointCloudRenderer::DrawOrder::TopToBottom );
```
Comment (nyalldawson) [inline src/gui/pointcloud/qgspointcloudrendererpropertieswidget.cpp:106]:
```suggestion
  mDrawOrderComboBox->addItem( tr( "Bottom to Top" ), QgsPointCloudRenderer::DrawOrder::BottomToTop );
  mDrawOrderComboBox->addItem( tr( "Top to Bottom" ), QgsPointCloudRenderer::DrawOrder::TopToBottom );
```

**Comment by nyalldawson** [comment]
Can you add some unit tests too please?

File: src/core/pointcloud/qgspointcloudlayerrenderer.cpp (hunk 1 above)
Comment (uclaros) [reply src/core/pointcloud/qgspointcloudlayerrenderer.cpp:122]: We only need to cater here for draw orders that require the Z value, ie bottom to top and top to bottom. Also combining it with `!context.renderContext().zRange().isInfinite() ||` will look a little ugly imho.

File: src/core/pointcloud/qgspointcloudlayerrenderer.cpp (hunk 2 above)
Comment (uclaros) [reply src/core/pointcloud/qgspointcloudlayerrenderer.cpp:201]: This is just indented existing code, but I could change that to a switch statement, yes.

Comment (nyalldawson) [reply src/core/pointcloud/qgspointcloudlayerrenderer.cpp:201]: yes please!

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**